bitnami/common/templates/_affinities.tpl

```diff
@@ -77,15 +78,29 @@ preferredDuringSchedulingIgnoredDuringExecution:
           {{- end }}
       topologyKey: {{ include "common.affinities.topologyKey" (dict "topologyKey" .topologyKey) }}
     weight: 1
+  {{- range $extraLabelSelectors }}
+  - podAffinityTerm:
+      labelSelector:
+        matchLabels: {{- (include "common.labels.matchLabels" $.context) | nindent 10 }}
+          {{- if not (empty $component) }}
+          {{ printf "app.kubernetes.io/component: %s" $component }}
+          {{- end }}
+          {{- range $key, $value := .extraMatchLabels }}
+          {{ $key }}: {{ $value | quote }}
+          {{- end }}
+      topologyKey: {{ include "common.affinities.topologyKey" (dict "topologyKey" .topologyKey) }}
+    weight: {{ .weight | default 1 -}}
+  {{- end -}}
```

Comment:
> my main goal is to add extraLabelSelectors to hard affinities

You can use the already existing `extraMatchLabels` parameter for that, see:

- https://github.com/bitnami/charts/blob/main/bitnami/common/templates/_affinities.tpl#L95-L97

Reply:
In that case I cannot combine 2 topologyKeys. I want to have the ability to generate 2 labelSelector entries: one for pushing to other zone, and another one to push to another node. Here is an example of desired affinity rules for my case.
`      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchLabels:
                app.kubernetes.io/component: clickhouse
                app.kubernetes.io/instance: clickhouse-bitnami
                app.kubernetes.io/name: clickhouse
            topologyKey: kubernetes.io/hostname
          - labelSelector:
              matchLabels:
                app.kubernetes.io/component: clickhouse
                app.kubernetes.io/instance: clickhouse-bitnami
                app.kubernetes.io/name: clickhouse
                shard: "0"
            topologyKey: topology.kubernetes.io/zone
`
I understand I could override entire affinity block, but in my case [Statefulset](https://github.com/bitnami/charts/blob/main/bitnami/clickhouse/templates/statefulset.yaml) is generated for each shard and affinity rules will be the same for each statefulset. To push pods of the same shard to another zone i need to point specific shard value for each statefulset. But shard value is not accessible from values because it is created on a template level by [iteration](https://github.com/bitnami/charts/blob/main/bitnami/clickhouse/templates/statefulset.yaml#L7C1-L7C37).
So I decided to add extraLabelsSelector with customizable topologyKey to have the ability to generate affinity rules on a template level. The rules above is generated by calling a template like this:
` podAntiAffinity: {{- include "common.affinities.pods" (dict "type" $.Values.podAntiAffinityPreset "component" "clickhouse" "extraLabelSelectors" (list (dict "extraMatchLabels" (dict "shard" $i) "topologyKey" "topology.kubernetes.io/zone")) "context" $) | nindent 10 }}
`

Reply:
I see your point...
In that case, we should definitely rename it to `extraPodAffinityTerms` to cover your use case since both "hard" and "soft" pod affinities expect an array of these objects (the only difference is that you can indicate the weight for each term in the "soft" type):
```console
$ kubectl explain sts.spec.template.spec.affinity.podAntiAffinity
GROUP:      apps
KIND:       StatefulSet
VERSION:    v1

FIELD: podAntiAffinity <PodAntiAffinity>

DESCRIPTION:
    Describes pod anti-affinity scheduling rules (e.g. avoid putting this pod in
    the same node, zone, etc. as some other pod(s)).
    Pod anti affinity is a group of inter pod anti affinity scheduling rules.

FIELDS:
  preferredDuringSchedulingIgnoredDuringExecution	<[]WeightedPodAffinityTerm>
    The scheduler will prefer to schedule pods to nodes that satisfy the
    anti-affinity expressions specified by this field, but it may choose a node
    that violates one or more of the expressions. The node that is most
    preferred is the one with the greatest sum of weights, i.e. for each node
    that meets all of the scheduling requirements (resource request,
    requiredDuringScheduling anti-affinity expressions, etc.), compute a sum by
    iterating through the elements of this field and adding "weight" to the sum
    if the node has pods which matches the corresponding podAffinityTerm; the
    node(s) with the highest sum are the most preferred.

  requiredDuringSchedulingIgnoredDuringExecution	<[]PodAffinityTerm>
    If the anti-affinity requirements specified by this field are not met at
    scheduling time, the pod will not be scheduled onto the node. If the
    anti-affinity requirements specified by this field cease to be met at some
    point during pod execution (e.g. due to a pod label update), the system may
    or may not try to eventually evict the pod from its node. When there are
    multiple elements, the lists of nodes corresponding to each podAffinityTerm
    are intersected, i.e. all terms must be satisfied.
```

Reply:
You mean I should keep the structure as it is now in my PR, but replace `extraLabelSelectors` with `extraPodAffinityTerms` for both `soft` and `hard` templates?